docs/python-api/python-semantics.rst

```diff
@@ -0,0 +1,36 @@
+Triton mostly follows the semantics of NumPy with minor exceptions. In this document, we go over some of the array computing features supported in Triton, and we cover the exceptions where Triton's semantics deviate from that NumPy.
+
+Type Promotion
+==============
+
+**Type Promotion** occurs when tensors of different data types are used in an operation. For binary operations associated to `dunder methods <https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types>`_ and the ternary function ``tl.where`` on its last two arguments, Triton automatically converts the input tensors to a common data type following a hierarchy of kinds (sets of dtypes): ``{bool} < {integral dypes} < {floating point dtypes}``.
+
+The algorithm is as follows:
+
+1. **Kind** If one tensor is of a dtype of a higher kind, the other tensor is promoted to this dtype: ``(int32, bfloat16) -> bfloat16``
+
+2. **Width** If both tensors are of dtypes of the same kind, and one of them is of a higher width, the other one is promoted to this dtype: ``(float32, float16) -> float32``
+
+3. **Supremum** If both tensors are of the same width and signedness but different dtypes, they are both promoted to the next larger dtype. ``(float16, bfloat16) -> float32``
+
+   3.1 If both tensors are of ``fp8`` dtypes, they are both casted to ``bfloat16``.
+
+4. **Prefer signed** Otherwise (same width, different signedness), they are promoted to the signed dtype: ``(int32, uint32) -> int32``
+
+When an operation involves a tensor and a scalar
+
+1. If the scalar is of a kind lower or equal to the tensor, it will not participate in the promotion: ``(float32, int) -> float32``
+
+2. If the scalar is of a higher kind, we choose the lowest dtype in which it fits among ``int32`` < ``uint32`` < ``int64`` < ``uint64`` for ints and ``float32`` < ``float64`` for floats. Then, both the tensor and the scalar are promoted to this dtype: ``(int16, 4.0) -> float32``
+
+
+Broadcasting
+============
+
+**Broadcasting** allows operations on tensors of different shapes by automatically expanding their shapes to a compatible size without copying the data. This follows the following rules:
+
+1. If one of the tensor shapes is shorter, pad it on the left with ones until both tensors have the same number of dimensions: ``((3, 4), (5, 3, 4)) -> ((1, 3, 4), (5, 3, 4))``
+
+2. Two dimensions are compatible if they are equal, or if one of them is 1. A dimension of 1 will be expanded to match the dimension of the other tensor. ``((5, 3, 4), (5, 3, 4)) -> ((5, 3, 4), (5, 3, 4))``
+
+Note that a dimension is expanded by creating a view on the original tensor with stride zero along that dimension. As such, we don't need to use any more memory for the expanded tensor.
```

Comment:
I don't think this is true in triton. In registers the compiler does try to de-duplicate elements within a single thread, but you can still have elements duplicated across threads. Also if the tensor is moved to shared memory for a tensor core operation, we really use the full sized shared memory.

Reply:
that's fair. Removed.